manual/src/tutorials/effectexamples.etex

```diff
@@ -0,0 +1,666 @@
+let run main =
+  let exchanger = ref None in (* waiting exchanger *)
+  let run_q = Queue.create () in (* scheduler queue *)
+  let enqueue k v =
+    Queue.push (fun _ -> continue k v) run_q
+  in
+  let dequeue () =
+    if Queue.is_empty run_q then () (* done *)
+    else (Queue.pop run_q) ()
+  in
+  let rec spawn f =
+    match_with f () {
+      retc = dequeue;
+      exnc = (fun e ->
+        print_endline (Printexc.to_string e);
+        dequeue ());
+      effc = fun (type a) (eff : a t) ->
+        match eff with
+        | Yield -> Some (fun (k : (a, unit) continuation) ->
+            enqueue k (); dequeue ())
+        | Fork f -> Some (fun (k : (a, unit) continuation) ->
+            enqueue k (); spawn f)
+        | Xchg n -> Some (fun (k : (int, unit) continuation) ->
+            begin match !exchanger with
+            | Some (n', k') ->
+                exchanger := None; enqueue k' n; continue k n'
+            | None -> exchanger := Some (n, k); dequeue ()
+            end)
+        | _ -> None
+    }
+  in
+  spawn main
+\end{caml_example*}
+
+We use a mutable queue "run_q" to hold the scheduler queue. "enqueue" pushes
+tasks into the queue, and "dequeue" pops tasks from the queue and runs them. We
+also have a ref cell "exchanger" that holds the task offering to exchange a
+value.
+
+The heavy lifting is done by the "spawn" function. The "spawn" function runs
+the given computation "f" in an effect handler. If "f" returns with a value
+(case "retc"), we dequeue and run the next task from the scheduler queue. If
+the computation "f" raises an exception (case "exnc"), we print the exception
+and run the next task from the scheduler.
+
+The computation "f" may also perform effects. If "f" performs the "Yield"
+effect, we enqueue the current task and run the next one from the scheduler
+queue. If the effect is "Fork f", we enqueue the current task, and recursively
+call "spawn" on "f", which run the given task in its own handler effectively
+running the task concurrently.
+
+If the effect is "Xchg", then we first check whether there is a task waiting to
+exchange. If so, we enqueue the waiting task with the current value being
+offered, and immediately resume the current task with the value being offered.
+If not, we make the current task the waiting exchanger, and run the next task
+from the scheduler queue.
+
+Now we can write a concurrent program that utilises the newly defined
+operations:
+
+\begin{caml_example}{verbatim}
+open Printf
+
+let _ = run (fun _ ->
+  fork (fun _ ->
+    printf "[t1] Sending 0\n";
+    printf "[t1] received %d\n" (xchg 0));
+  fork (fun _ ->
+    printf "[t2] Sending 1\n";
+    printf "[t2] received %d\n" (xchg 1)))
+\end{caml_example}
+
+Observe that the messages from the two tasks are interleaved. Notice also that
+the snippet above makes no reference to the effect handlers, and is in
+direct-style (no monadic operations). This example illustrates that, with
```

Comment:
Will the terse mentions of monads be understandable or helpful to the intended audience?

Reply:
Quite difficult to tersely explain a concurrency monad. :-/

Reply:
If someone has worked with concurrency monads, they probably can notice the difference by themselves. (So that sentence could be omitted.)

An alternative is to say something more concrete like "no continuation-passing" or "no building continuations as closures by the programmer". This should apply to both strict CPS and monadic style.